File: go/core/errors_test.go
// Copyright (c) 2026 Omas Cloud
//
// SPDX-License-Identifier: MIT

package core

import (
	"context"
	"errors"
	"io"
	"net/http"
	"strings"
	"testing"
	"time"
)

func TestDecodeAPIErrorPreservesProtocolDetails(t *testing.T) {
	response := &http.Response{
		StatusCode: http.StatusNotFound,
		Header: http.Header{
			"X-Request-Id": {"request-123"},
			"Retry-After":  {"17"},
		},
		Body: io.NopCloser(strings.NewReader("unused")),
	}
	body := []byte(`{"error":"missing metric","errorCode":"RESOURCE_NOT_FOUND","extraData":{"resourceType":"metric"}}`)

	apiError := decodeAPIError(response, body)

	if apiError.StatusCode != http.StatusNotFound || apiError.ErrorCode != "RESOURCE_NOT_FOUND" ||
		apiError.Message != "missing metric" || apiError.RequestID != "request-123" ||
		apiError.RetryAfter != 17*time.Second {
		t.Fatalf("unexpected API error: %+v", apiError)
	}
	if string(apiError.ExtraData) != `{"resourceType":"metric"}` || string(apiError.RawBody) != string(body) {
		t.Fatalf("response data not preserved: %+v", apiError)
	}
	response.Header.Set("X-Request-Id", "mutated")
	body[0] = 'x'
	if apiError.Headers.Get("X-Request-Id") != "request-123" || apiError.RawBody[0] != '{' {
		t.Fatal("API error retained mutable response storage")
	}
	if apiError.Error() != "RESOURCE_NOT_FOUND: missing metric" {
		t.Fatalf("error string = %q", apiError.Error())
	}
}

func TestDecodeAPIErrorFallsBackForMalformedBody(t *testing.T) {
	response := &http.Response{StatusCode: http.StatusBadGateway, Header: http.Header{}}
	apiError := decodeAPIError(response, []byte("not-json"))
	if apiError.Message != "API request failed with HTTP 502" || apiError.ErrorCode != "" {
		t.Fatalf("unexpected fallback: %+v", apiError)
	}
}

func TestRuntimeErrorsUnwrapCauses(t *testing.T) {
	cause := context.DeadlineExceeded
	errorsToCheck := []error{
		&AuthenticationError{OperationID: "ListMetrics", Err: cause},
		&RequestTimeoutError{OperationID: "ListMetrics", Err: cause},
		&SerializationError{OperationID: "ListMetrics", Err: cause},
		&TransportError{OperationID: "ListMetrics", Err: cause},
	}
	for _, sdkError := range errorsToCheck {
		if !errors.Is(sdkError, cause) {
			t.Fatalf("%T does not unwrap its cause", sdkError)
		}
		if sdkError.Error() == "" {
			t.Fatalf("%T has an empty message", sdkError)
		}
	}
}

File: typescript/core/src/errors.test.ts
import { describe, expect, it } from "vitest";

import {
    ApiError,
    AuthenticationError,
    RequestTimeoutError,
    SdkError,
    SerializationError,
    TransportError,
} from "./errors.js";

describe("ApiError", () => {
    it("retains safe protocol metadata and typed details", () => {
        const headers = new Headers({ "x-request-id": "request-1" });
        const error = new ApiError({
            status: 404,
            errorCode: "RESOURCE_NOT_FOUND",
            message: "missing",
            requestId: "request-1",
            retryAfterMs: 2_000,
            headers,
            details: { resourceType: "metric" },
        });
        headers.set("x-request-id", "changed");

        expect(error).toBeInstanceOf(SdkError);
        expect(error.name).toBe("ApiError");
        expect(error.message).toBe("RESOURCE_NOT_FOUND: missing");
        expect(error.serverMessage).toBe("missing");
        expect(error.status).toBe(404);
        expect(error.errorCode).toBe("RESOURCE_NOT_FOUND");
        expect(error.requestId).toBe("request-1");
        expect(error.retryAfterMs).toBe(2_000);
        expect(error.headers.get("x-request-id")).toBe("request-1");
        expect(error.details).toEqual({ resourceType: "metric" });
        expect(error).not.toHaveProperty("rawBody");
    });

    it("uses the server message when no stable code is available", () => {
        const error = new ApiError({
            status: 500,
            errorCode: "",
            message: "API request failed with HTTP 500",
            headers: new Headers(),
        });

        expect(error.message).toBe("API request failed with HTTP 500");
    });
});

describe("runtime errors", () => {
    it.each([
        [AuthenticationError, "AuthenticationError"],
        [TransportError, "TransportError"],
        [RequestTimeoutError, "RequestTimeoutError"],
        [SerializationError, "SerializationError"],
    ] as const)("%s preserves its cause", (ErrorType, name) => {
        const cause = new Error("safe cause");
        const error = new ErrorType("listMetrics", "request failed", cause);

        expect(error).toBeInstanceOf(SdkError);
        expect(error.name).toBe(name);
        expect(error.cause).toBe(cause);
        expect(error.message).toBe("request failed");
    });
});
